dpbench/infrastructure/benchmark_runner.py

```diff
@@ -0,0 +1,403 @@
+from dpbench.infrastructure.frameworks import Framework
+from dpbench.infrastructure.frameworks.fabric import build_framework
+from dpbench.infrastructure.runner import _print_results
+
+"""
+Send on process creation:
+    Framework configuration: so the process can set it up.
+
+Send on benchmark run request:
+    Benchmark configuration
+
+Send on benchmark finish:
+    Benchmark execution results
+"""
+
+
+@dataclass
+class BaseRunConfig:
+    """Run configuration that is required to run benchmark."""
+
+    benchmark: cfg.Benchmark
+    implementation: str
+    ref_framework: cfg.Framework = None
+    preset: str = "S"
+    repeat: int = 10
+    validate: bool = True
+    precision: str = None
+    print_results: bool = True
+
+    @classmethod
+    def from_instance(cls, instance):
+        """Creates BaseRunConfiguration from it's inheritor."""
+        return cls(
+            **{
+                k: v
+                for k, v in instance.__dict__.items()
+                if k in inspect.signature(cls).parameters
+            }
+        )
+
+
+@dataclass
+class RunConfig(BaseRunConfig):
+    """Extended run configuration that controls benchmark execution."""
+
+    conn: sqlalchemy.Engine = None
+    framework: cfg.Framework = None
+    timeout: float = 200.0
+    run_id: int = None
+    skip_expected_failures: bool = False
+
+
+class BenchmarkRunner:
+    """Benchmark runner that delegates runs between processes.
+
+    It creates process for each framework to avoid conflicts in framework
+    imports.
+    """
+
+    def __init__(
+        self,
+        method: str = "spawn",
+    ) -> None:
+        """Creates BenchmarkRunner. No processes get spawn at this point.
+
+        Args:
+            method: method for sub process creations. Currently only spawn is
+                supported. We need to get rid of GLOBAL config to get support
+                for 'fork' method.
+        """
+        self._ctx = mp.get_context(method)
+        self._framework_processes: dict[
+            str, tuple[mp.Process, mpc.Connection]
+        ] = {}
+
+    def get_process(
+        self, framework: cfg.Framework
+    ) -> tuple[mp.Process, mpc.Connection]:
+        """Get process with connection to it using caching.
+
+        Caching is working the way that if the process does not exists it will
+        be created automatically. Framework comparison is done by
+        framework.simple_name key.
+
+        Args:
+            framework: framework to which return the process.
+
+        Returns: tuple of process and connection pip to control it.
+        """
+        p, conn = self._framework_processes.get(
+            framework.simple_name, (None, None)
+        )
+        if not p:
+            return self.create_process(framework)
+        return p, conn
+
+    def kill_process(self, framework: cfg.Framework) -> None:
+        """Kill the process for framework and closes connection.
+
+        Args:
```

Comment:
General comment about docstring, does pydocstyle not flag the missing `Returns` tag?

Reply:
Nope. There is an issue with pydocstyle detecting google style comments

Reply:
I guess you mentioned few PRs ago, that `Returns: nothing.` causes issues with sphinx?

Reply:
Ya, we have to generate Sphinx and see. We will address these once the API docs generation is setup.